okhttp/src/main/java/okhttp3/WebSocketListener.java

```diff
@@ -18,27 +18,34 @@
 import java.io.IOException;
 import okio.Buffer;
 
-/** Listener for server-initiated messages on a connected {@link WebSocket}. */
+/**
+ * Listener for server-initiated messages on a connected {@link WebSocket}. All callbacks will be
+ * called on a single thread.
+ *
+ * <h2>Lifecycle Rules</h2>
+ * <ul>
+ * <li>Either {@link #onOpen} or {@link #onFailure} will be called first depending on if the web
+ * socket was successfully opened or if there was an error connecting to the server or parsing its
+ * response.</li>
+ * <li>After {@link #onOpen} is called, {@link #onFailure} can be called at any time. No more
+ * callbacks will follow a call to {@link #onFailure}.</li>
+ * <li>After {@link #onOpen} is called, {@link #onMessage} and {@link #onPong} will be called for
+ * each message and pong frame, respectively. Note: {@link #onPong} may be called while {@link
```

Comment:
Yep! Not sure what to do otherwise. A queue would skew timings if you were using ping/pongs to measure.